build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,255 @@
+  }
+
+  if (t.isBinaryExpression(node, {operator: '+'})) {
+    return (
+      buildMessage(t, node.left, interpolationArgs) +
+      buildMessage(t, node.right, interpolationArgs)
+    );
+  }
+
+  if (t.isTemplateLiteral(node)) {
+    let quasied = '';
+    let i = 0;
+    for (; i < node.quasis.length - 1; i++) {
+      quasied += node.quasis[i].value.cooked;
+      quasied += buildMessage(t, node.expressions[i], interpolationArgs);
+    }
+    quasied += node.quasis[i].value.cooked;
+    return quasied;
+  }
+
+  interpolationArgs.push(node);
+  return '%s';
+}
+
+module.exports = function({types: t}) {
+  return {
+    pre() {
+      // Configurable to isolate test output.
+      const {messagesPath = defaultMessagesPath} = this.opts;
+      this.messagesPath = relativeToRoot(messagesPath);
+    },
+    visitor: {
+      CallExpression({node}, {opts}) {
+        const {callee} = node;
+
+        // Looks like a method().call().
+        if (
+          !t.isMemberExpression(callee) ||
+          !t.isCallExpression(callee.object)
+        ) {
+          return;
+        }
+
+        const singletonCallee = callee.object.callee;
+        if (
+          !singletonFunctions.some(name =>
+            t.isIdentifier(singletonCallee, {name})
+          )
+        ) {
+          return;
+        }
+
+        const {property} = callee;
+        if (
+          !t.isIdentifier(property) ||
+          !(property.name in transformableMethods)
+        ) {
+          return;
+        }
+
+        const {messageArgPos} = transformableMethods[property.name];
+        const messageArg = node.arguments[messageArgPos];
+
+        if (!messageArg) {
+          return;
+        }
+
+        // Construct a printf template from the argument set. There could be
+        // non-string types among string literals in variadic calls, so the
+        // template includes them as interpolated arguments.
+        const templateArgs = [];
+        const message = buildMessage(t, messageArg, templateArgs);
+
+        // Bounce when indirection increases minified size (±1 byte). Also
+        // catches the the case where the top-level message is variable (ie.
+        // when its template is exactly '%s'), in which indirection would be
+        // pointless.
+        if (message.length <= roughMinifiedExpansionLength) {
+          return;
+        }
+
+        const shortMessageId = getOrCreateShortMessageId(
+            this.messagesPath,
+            message
+        );
+
+        // Temporary option to not replace call arguments, but still output the
+        // table to keep other code and infra independent from rollout.
+        if (opts.replaceCallArguments === false) {
+          return;
+        }
+
+        // We care only about message arguments beyond the first (ie. the body):
+        // all previous method arguments (ie. the head) should stay in place.
+        // The body is nested into the expansion method call, after the template
+        // id.
```

Comment:
is this because head is usually the `TAG`?

Reply:
Yeah, anything before the first message arg is part of the higher-level call.

Maybe we can enforce the usage of `TAG` some other way, so that it's not actually included in the top-level call but rather in the message template string. A few more bytes squeezed out.

Reply:
yeah, i feel like that `TAG` argument is incorrectly used as well. we can defer, but ideally you are right in that it should be part of the message

Reply:
Let's defer so we can come up with a unified solution that involves groundwork for (internal) http://go/amp-formatted-logging. This should involve deprecating variadic calls in favor of template literals (suggested by @kristoferbaxter) and inclusion of `TAG`. For now I think this PR reduces enough.